### PR TITLE
[docs][1.5] Update distributed autograd note

### PR DESCRIPTION
**Pull request by rohan-varma**
- Update API calls `backward` and `optim.step` now that we require `context_id`
- Add notes to clarify purpose of distributed autograd context (this was a source of confusion in some feedback)
- Add note that details why optimizer requires context_id
- Clearly specify that we don't have SMART mode yet

### CODE REVIEW

File: docs/source/notes/distributed_autograd.rst [hunk 1]
```diff
@@ -113,7 +113,12 @@ From the user's perspective the autograd context is setup as follows:
   import torch.distributed.autograd as dist_autograd
   with dist_autograd.context() as context_id:
     loss = model.forward()
-    dist_autograd.backward(loss)
+    dist_autograd.backward(context_id, loss)
```
Comment (mrshenli) [inline docs/source/notes/distributed_autograd.rst]: is it mandatory to specify the `context_id` here, or will it infer the id if not provided?

File: docs/source/notes/distributed_autograd.rst [hunk 2]
```diff
@@ -307,7 +315,11 @@ The :class:`~torch.distributed.optim.DistributedOptimizer` operates as follows:
    each of the worker nodes and holds an ``RRef`` to them.
 4. When :meth:`torch.distributed.optim.DistributedOptimizer.step` is invoked,
    the distributed optimizer uses RPC to remotely execute all the local
-   optimizers on the appropriate remote workers.
+   optimizers on the appropriate remote workers. A distributed autograd 
+   `context_id` must be provided as input to
```
Comment (mrshenli) [inline docs/source/notes/distributed_autograd.rst]: Would I be correct if I assume we used to infer the id, but now we no longer support that?

File: docs/source/notes/distributed_autograd.rst (hunk 1 above)
Comment (rohan-varma) [reply docs/source/notes/distributed_autograd.rst]: Yep, the context_id is required now: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/init.cpp

File: docs/source/notes/distributed_autograd.rst (hunk 2 above)
Comment (rohan-varma) [reply docs/source/notes/distributed_autograd.rst]: Yes, the context_id is required now: https://github.com/pytorch/pytorch/blob/master/torch/distributed/optim/optimizer.py#L125

This was part of the change in https://github.com/pytorch/pytorch/issues/33480

**Comment by rohan-varma** [comment]
Changes: 

change_1
change_2
change_3

File: docs/source/notes/distributed_autograd.rst (hunk 2 above)
Comment (osalpekar) [reply docs/source/notes/distributed_autograd.rst]: @mrshenli Here is the closed PR for this change: #33711. We used to the depend on the thread_local `context_id` in the autograd container before, and internally call `currentContext`. Now the context_id is explicitly passed in and we internally call `retrieveContext(context_id)`. dist_optim.step() also takes in the context_id now.

File: docs/source/notes/distributed_autograd.rst [hunk 3]
```diff
@@ -113,7 +113,12 @@ From the user's perspective the autograd context is setup as follows:
   import torch.distributed.autograd as dist_autograd
   with dist_autograd.context() as context_id:
     loss = model.forward()
-    dist_autograd.backward(loss)
+    dist_autograd.backward(context_id, loss)
+
+It is important to note that your model's forward pass must be invoked within
+the distributed autograd context manager, as a valid context is needed in
+order to ensure that all ``send`` and ``recv`` functions are stored properly
+to run the backwards pass across all participating nodes.
```
Comment (osalpekar) [inline docs/source/notes/distributed_autograd.rst]: nit: backward pass

**Review by osalpekar — COMMENTED** [review]
Looks good, thanks @rohan-varma ! I'll wait for @pritamdamania87  or @mrshenli to do a final stamp

File: docs/source/notes/distributed_autograd.rst (hunk 2 above)
Comment (mrshenli) [inline docs/source/notes/distributed_autograd.rst]: Shall we use double quote here for code?

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks!

**Pull request closed without merging**